web-client/src/HabitUtil.elm

```diff
@@ -0,0 +1,188 @@
+module HabitUtil exposing (..)
+
+{-| Module for useful Habit operations
+-}
+
+import Models.Habit as Habit
+import Models.FrequencyStats as FrequencyStats
+
+
+isHabitCurrentFragmentSuccessful : Habit.Habit -> FrequencyStats.FrequencyStats -> Bool
+isHabitCurrentFragmentSuccessful habit frequencyStats =
+    case habit of
+        Habit.GoodHabit _ ->
+            frequencyStats.currentFragmentTotal >= frequencyStats.currentFragmentGoal
+
+        Habit.BadHabit _ ->
+            frequencyStats.currentFragmentTotal <= frequencyStats.currentFragmentGoal
+
+
+findFrequencyStatsForHabit : Habit.Habit -> List FrequencyStats.FrequencyStats -> Result String FrequencyStats.FrequencyStats
+findFrequencyStatsForHabit habit frequencyStats =
+    let
+        habitId =
+            (.id (Habit.getCommonFields habit))
+
+        filteredFrequencyStatsByHabitId =
+            List.filter (\stats -> stats.habitId == habitId) frequencyStats
+    in
+        case filteredFrequencyStatsByHabitId of
+            [] ->
+                Err ("No frequency stats found for habit " ++ habitId)
+
+            [ s ] ->
+                Ok s
+
+            s1 :: s2 :: _ ->
+                Err ("More than one frequency stats found for habit " ++ habitId ++ ", something fishy is going on")
+
+
+{-| Returns the habits sorted by completion, urgency, and current goal progress.
+-}
+sortHabitsByCurrentFragment : List FrequencyStats.FrequencyStats -> List Habit.Habit -> List Habit.Habit
```

Comment:
Use a sort function and just pass it comparators, you can see this here:

https://github.com/amilner42/code-tidbit/blob/a8a8b3b169eb7a4c929788095ae43353d6559752/frontend/src/DefaultServices/Sort.elm#L24

Also don't put the comparators inside the function, have them be their own functions, functions with huge `let` become hard to read